### PR TITLE
balancer: Add tracing dynamic config parameters

### PR DESCRIPTION
**Pull request by jkosh44**
This commit adds dynamic configuration parameters to balancerd that can
control tracing.

### Motivation
This PR adds a feature that has not yet been specified.

### Checklist

- [X] This PR has adequate test coverage / QA involvement has been duly considered. ([trigger-ci for additional test/nightly runs](https://trigger-ci.dev.materialize.com/))
- [X] This PR has an associated up-to-date [design doc](https://github.com/MaterializeInc/materialize/blob/main/doc/developer/design/README.md), is a design doc ([template](https://github.com/MaterializeInc/materialize/blob/main/doc/developer/design/00000000_template.md)), or is sufficiently small to not require a design.
  
- [X] If this PR evolves [an existing `$T ⇔ Proto$T` mapping](https://github.com/MaterializeInc/materialize/blob/main/doc/developer/command-and-response-binary-encoding.md) (possibly in a backwards-incompatible way), then it is tagged with a `T-proto` label.
- [X] If this PR will require changes to cloud orchestration or tests, there is a companion cloud PR to account for those changes that is tagged with the release-blocker label ([example](https://github.com/MaterializeInc/cloud/pull/5021)).
  
- [X] If this PR includes major [user-facing behavior changes](https://github.com/MaterializeInc/materialize/blob/main/doc/developer/guide-changes.md#what-changes-require-a-release-note), I have pinged the relevant PM to schedule a changelog post.

### CODE REVIEW

File: src/balancerd/src/lib.rs [hunk 1]
```diff
@@ -178,6 +182,10 @@ impl BalancerService {
         let metrics = BalancerMetrics::new(&cfg);
         let mut configs = ConfigSet::default();
         configs = dyncfgs::all_dyncfgs(configs);
+        match tracing_config(&configs) {
+            Ok(parameters) => parameters.apply(&cfg.tracing_handle),
+            Err(e) => warn!("unable to update tracing: {e}"),
+        }
```
Comment (jkosh44) [inline src/balancerd/src/lib.rs:188]: TODO(jkosh44) plumb this to dyncfg sync.

Comment (benesch) [reply src/balancerd/src/lib.rs:188]: You probably don't even need it here if you have it in the sync, right?

Are you thinking of adding another callback to `sync_launchdarkly_to_configset` that is invoked on every sync?

**Review by benesch — COMMENTED** [review]
Looking good!

File: src/balancerd/src/lib.rs (hunk 1 above)
Comment (jkosh44) [reply src/balancerd/src/lib.rs:188]: Yes that's exactly what I was thinking (and just did).

File: src/balancerd/src/lib.rs [hunk 2]
```diff
@@ -209,6 +217,23 @@ impl BalancerService {
...
+            move |updates, configs| {
+                if [
+                    LOGGING_FILTER.name(),
+                    OPENTELEMETRY_FILTER.name(),
+                    LOGGING_FILTER_DEFAULTS.name(),
+                    OPENTELEMETRY_FILTER_DEFAULTS.name(),
+                    SENTRY_FILTERS.name(),
+                ]
+                .into_iter()
+                .any(|name| updates.updates.contains_key(name))
```
Comment (benesch) [inline src/balancerd/src/lib.rs:229]: Seems worth pulling out a `has_tracing_config_change` function and sticking it next to `tracing_config`, just to increase the odds that if you change one you remember to change the other.

**Review by benesch — APPROVED** [review]
Lovely, thank you!

**Pull request merged**